portal/models/i18n_utils.py

```diff
@@ -302,3 +308,177 @@ def compile_pos():
             "Saved MO file: %s",
             os.path.relpath(mo_filepath, current_app.root_path),
         )
+
+
+def upsert_to_template_file():
+    db_translatables = {}
+    db_translatables.update(get_db_strings())
+    if not db_translatables:
+        current_app.logger.warn("no DB strings extracted")
+        return
+
+    db_translatables.update(get_static_strings())
+
+    try:
+        with open(
+            os.path.join(
+                current_app.root_path,
+                "translations/messages.pot",
+            ),
+            "r+",
+        ) as potfile:
+            potlines = potfile.readlines()
+            for i, line in enumerate(potlines):
+                if not line.split() or (line.split()[0] != "msgid"):
+                    continue
+                msgid = line.split(" ", 1)[1].strip()
+                if msgid not in db_translatables:
+                    continue
+                for location in db_translatables[msgid]:
+                    locstring = "# " + location + "\n"
+                    if not any(t == locstring for t in potlines[i - 4:i]):
+                        potlines.insert(i, locstring)
+                del db_translatables[msgid]
+            for entry, locations in db_translatables.items():
+                if not entry:
+                    continue
+                for loc in locations:
+                    potlines.append("# " + loc + "\n")
+                potlines.append("msgid " + entry + "\n")
+                potlines.append("msgstr \"\"\n")
+                potlines.append("\n")
+            potfile.truncate(0)
+            potfile.seek(0)
+            potfile.writelines(potlines)
+    except:
+        exceptionType, exceptionValue, exceptionTraceback = sys.exc_info()
+        sys.exit(
+            "Could not write to translation file!\n ->%s" % (exceptionValue))
+
+
+def fix_references(pot_fpath):
+    """Fix reference comments to remove checkout-specific paths"""
+    # Todo: override PoFileParser._process_comment() to perform this as part of `pybabel extract`
+
+    path_regex = re.compile(r"^#: {}(?P<rel_path>.*):(?P<line>\d+)".format(
```

Comment:
possibly a premature optimization, but why re-compile the regex every time in the loop?

Reply:
Not ideal, but I can't think of a good way to keep that function self-contained and not re-compile the regex every time.

Some kind of memoization to cache it? But I think I would need a separate function just to build the regex?